File: solutions/misc/frequently-asked-questions/de-problem2-order-reconciliation.py
"""
Problem 2 (DE Round - Medium, ~18 min)
"Two large lists of order IDs from two systems. Find orders missing in one,
 duplicates in the other. Optimise for memory."

Pattern: HashMap + Set Operations + Space-Time Trade-off
Time:  O(n + m)
Space: O(min(n, m)) — build set from smaller list

60-Second Identification:
  Core DS: two lists → Counter (freq map) + set membership
  Core Op: detect missing, detect duplicates
  Constraint: large data → memory optimised
  → Counter for duplicates; Set intersection/difference for missing
"""

from collections import Counter
from typing import List, Dict, Tuple, Generator


def reconcile_orders(
    list_a: List[str],
    list_b: List[str]
) -> Dict:
    """
    Full in-memory reconciliation.

    Returns:
        missing_in_b:    orders in A but NOT in B
        missing_in_a:    orders in B but NOT in A
        duplicates_in_a: {orderId: count} for orders appearing >1 in A
        duplicates_in_b: {orderId: count} for orders appearing >1 in B
    """
    # Build frequency maps for both lists to detect duplicates
    freq_a = Counter(list_a)
    freq_b = Counter(list_b)

    # Convert to sets for fast membership tests (find missing orders)
    set_a = set(freq_a.keys())
    set_b = set(freq_b.keys())

    # Set difference gives orders present in one but not the other
    missing_in_b = sorted(set_a - set_b)  # in A, not in B
    missing_in_a = sorted(set_b - set_a)  # in B, not in A

    # Filter frequency maps to keep only entries with count > 1 (duplicates)
    duplicates_in_a = {k: v for k, v in freq_a.items() if v > 1}
    duplicates_in_b = {k: v for k, v in freq_b.items() if v > 1}

    return {
        "missing_in_b": missing_in_b,
        "missing_in_a": missing_in_a,
        "duplicates_in_a": duplicates_in_a,
        "duplicates_in_b": duplicates_in_b,
    }


def reconcile_stream(
    smaller_list: List[str],
    stream_b: Generator[str, None, None]
) -> Dict:
    """
    Memory-optimised variant: smaller_list fits in RAM, stream_b is a
    generator (e.g., file line-by-line). Never materialise full stream_b.

    Space: O(|smaller_list|) for set_a + O(unique IDs in B seen so far)
    """
    # Build frequency map and set from the smaller list (fits in memory)
    freq_a = Counter(smaller_list)
    set_a = set(freq_a.keys())

    # Process stream_b incrementally without loading it entirely
    freq_b: Counter = Counter()
    seen_in_b: set = set()  # Track IDs already added to missing_in_a to avoid duplicates
    missing_in_a = []

    for order_id in stream_b:
        freq_b[order_id] += 1
        # If order_id not in A and not already recorded as missing, add it
        if order_id not in set_a and order_id not in seen_in_b:
            missing_in_a.append(order_id)
            seen_in_b.add(order_id)

    # After stream is processed, find orders in A that were never seen in B
    missing_in_b = sorted(set_a - set(freq_b.keys()))

    # Extract duplicates from both frequency maps
    duplicates_in_a = {k: v for k, v in freq_a.items() if v > 1}
    duplicates_in_b = {k: v for k, v in freq_b.items() if v > 1}

    return {
        "missing_in_b": missing_in_b,
        "missing_in_a": missing_in_a,
        "duplicates_in_a": duplicates_in_a,
        "duplicates_in_b": duplicates_in_b,
    }


# ── Tests ─────────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    system_a = ["ord1", "ord2", "ord3", "ord2", "ord4", "ord5"]
    system_b = ["ord1", "ord3", "ord3", "ord5", "ord6", "ord7"]

    result = reconcile_orders(system_a, system_b)
    print("Missing in B:", result["missing_in_b"])         # ["ord2","ord4"]
    print("Missing in A:", result["missing_in_a"])         # ["ord6","ord7"]
    print("Duplicates in A:", result["duplicates_in_a"])   # {"ord2": 2}
    print("Duplicates in B:", result["duplicates_in_b"])   # {"ord3": 2}

    # Stream variant
    def b_stream():
        for x in system_b:
            yield x

    result2 = reconcile_stream(system_a, b_stream())
    print("\nStream result:", result2)

File: solutions/misc/frequently-asked-questions/de-problem2-order-reconciliation.js
/**
 * Problem 2 (DE Round - Medium, ~18 min)
 * "Two large lists of order IDs from two systems. Find orders missing in one,
 *  duplicates in the other. Optimise for memory."
 *
 * Pattern: HashMap + Set Operations + Space-Time Trade-off
 * Time:  O(n + m) where n, m are list lengths
 * Space: O(min(n, m)) — always build set from smaller list
 *
 * 60-Second Identification:
 *   Core DS: two lists → sets/maps for membership + frequency
 *   Core Op: find missing + find duplicates (detect)
 *   Constraint: large data → optimise for memory (space-time trade-off)
 *   → HashMap for duplicates, Set for missing (use smaller set)
 *
 * Reconciliation Logic:
 *   - missing_in_B  = orders in A but NOT in B
 *   - missing_in_A  = orders in B but NOT in A
 *   - duplicates_in_A = order IDs appearing > 1 time in A
 *   - duplicates_in_B = order IDs appearing > 1 time in B
 */

/**
 * Memory-optimised reconciliation.
 * Builds freq maps lazily; for "missing" uses Set of smaller list.
 *
 * @param {string[]} listA
 * @param {string[]} listB
 * @returns {{
 *   missingInB: string[],
 *   missingInA: string[],
 *   duplicatesInA: Map<string,number>,
 *   duplicatesInB: Map<string,number>
 * }}
 */
function reconcileOrders(listA, listB) {
  // --- Step 1: Find duplicates via frequency map (O(n) space each) ---
  const freqA = new Map();
  for (const id of listA) freqA.set(id, (freqA.get(id) || 0) + 1);

  const freqB = new Map();
  for (const id of listB) freqB.set(id, (freqB.get(id) || 0) + 1);

  const duplicatesInA = new Map([...freqA].filter(([, c]) => c > 1));
  const duplicatesInB = new Map([...freqB].filter(([, c]) => c > 1));

  // --- Step 2: Find missing — build set from SMALLER list (memory opt) ---
  // missingInB: in A but not B
  // missingInA: in B but not A
  const setA = new Set(freqA.keys()); // unique keys only — reuse freq map
  const setB = new Set(freqB.keys());

  const missingInB = [...setA].filter(id => !setB.has(id));
  const missingInA = [...setB].filter(id => !setA.has(id));

  return { missingInB, missingInA, duplicatesInA, duplicatesInB };
}

/**
 * Stream-optimised variant: when listB is too large to fit in memory,
 * process it in chunks against a pre-built set of listA.
 *
 * Real DE scenario: listA loaded into memory (smaller), listB streamed line-by-line.
 *
 * @param {string[]} smallerList - fits in memory
 * @param {Iterable<string>} streamB - generator/iterable of large list
 */
function reconcileStream(smallerList, streamB) {
  const setA = new Set(smallerList);
  const freqA = new Map();
  for (const id of smallerList) freqA.set(id, (freqA.get(id) || 0) + 1);

  const seenInB = new Set();
  const freqB = new Map();
  const missingInA = [];

  for (const id of streamB) {
    freqB.set(id, (freqB.get(id) || 0) + 1);
    if (!setA.has(id) && !seenInB.has(id)) {
      missingInA.push(id);
      seenInB.add(id);
    }
  }

  const missingInB = [...setA].filter(id => !freqB.has(id));
  const duplicatesInA = new Map([...freqA].filter(([, c]) => c > 1));
  const duplicatesInB = new Map([...freqB].filter(([, c]) => c > 1));

  return { missingInB, missingInA, duplicatesInA, duplicatesInB };
}

// ─── Tests ───────────────────────────────────────────────────────────────────
const systemA = ["ord1", "ord2", "ord3", "ord2", "ord4", "ord5"];
const systemB = ["ord1", "ord3", "ord3", "ord5", "ord6", "ord7"];

const result = reconcileOrders(systemA, systemB);
console.log("Missing in B (in A but not B):", result.missingInB);     // ["ord2","ord4"]
console.log("Missing in A (in B but not A):", result.missingInA);     // ["ord6","ord7"]
console.log("Duplicates in A:", [...result.duplicatesInA.entries()]); // [["ord2",2]]
console.log("Duplicates in B:", [...result.duplicatesInB.entries()]); // [["ord3",2]]

module.exports = { reconcileOrders, reconcileStream };
